### ISSUE TITLE
How to get notified when client close

### ISSUE THREAD

**Issue by 4eUeP** [Tradias/asio-grpc#51]
Suppose I have a server streaming rpc:

```
rpc ServerStream(Req) returns (stream Resp);
```

When a client calls ServerStream, the server does some bookkeeping; when the client disconnects, the bookkeeping needs to be removed. Is there an API, let's say `on_recv_client_close(release_function)`, that can register a callback for a client closed event? 

Thank you.

P.S.

- I know I can use failed state of the `agrpc::write` to indicate that client is closed. But I want to get notified even when the server doesn't send anything.
- grpc core has a `GRPC_OP_RECV_CLOSE_ON_SERVER` op that I don't know whether it helps.
- Here is a similar question I found but for grpc-go: https://stackoverflow.com/questions/39825671/grpc-go-how-to-know-in-server-side-when-client-closes-the-connection

**Comment by Tradias** [comment]
Yes that is possible through [AsyncNotifyWhenDone](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_context.html#ad51d67a4bd5a0960b4c15783e10a72a3), see also this [test](https://github.com/Tradias/asio-grpc/blob/master/test/src/test_high_level_client_17.cpp#L298-L348).

I can give you more details and examples on the weekend. Also thanks for reminding me about this gRPC feature, I will make sure to include support for it in the upcoming [high-level server API](https://github.com/Tradias/asio-grpc/issues/45).

**Comment by Tradias** [comment]
Since `agrpc::repeatedly_request` does not support pre-request interaction with the `grpc::ServerContext` you have to roll your own request_loop. Here is what that can look like:

```c++
template <class CompletionToken>
auto async_notify_when_done(agrpc::GrpcContext& grpc_context, grpc::ServerContext& server_context,
                            CompletionToken&& token)
{
    return agrpc::grpc_initiate(
        [&](const agrpc::GrpcContext&, void* tag)
        {
            server_context.AsyncNotifyWhenDone(tag);
        },
        asio::bind_executor(grpc_context, std::forward<CompletionToken>(token)));
}

template <class RequestHandler>
asio::awaitable<void> request_loop(agrpc::GrpcContext& grpc_context, test::v1::Test::AsyncService& service,
                                   RequestHandler request_handler)
{
    while (true)
    {
        grpc::ServerContext server_context;
        auto on_done = async_notify_when_done(grpc_context, server_context, asio::experimental::use_promise);
        test::msg::Request request;
        grpc::ServerAsyncWriter<test::msg::Response> writer{&server_context};
        const bool ok = co_await agrpc::request(&test::v1::Test::AsyncService::RequestServerStreaming, service,
                                                server_context, request, writer); // to improve performance you could bind a example::Buffer<> here
        if (!ok)
        {
            co_return;
        }
        asio::co_spawn(grpc_context, request_loop(grpc_context, service, request_handler), asio::detached);
        co_await request_handler(server_context, request, writer, std::move(on_done));
    }
}
```

Now to register your request handler:

```c++
auto request_handler = [&](grpc::ServerContext& server_context, test::msg::Request& request,
                            grpc::ServerAsyncWriter<test::msg::Response>& writer,
                            asio::experimental::promise<void(bool)> on_done) -> asio::awaitable<void>
{
    // Already cancelled?
    // const bool is_cancelled = on_done.complete();

    on_done.async_wait(
        [](bool)
        {
            // cancelled or finished
        });
};

// register the request handler
asio::co_spawn(grpc_context, request_loop(grpc_context, service, std::move(request_handler)), asio::detached);
```

Note that due to https://github.com/grpc/grpc/issues/10136, the GrpcContext will never run out of work after server shutdown, which also leads to a memory leak. I think the memory leak can be resolved on asio-grpc side by keeping track of all AsyncNotifyWhenDone completion handlers, such that they can be deallocated after CompletionQueue destruction. This will be part of the upcoming high-level server API. I might be able to provide a standalone `agrpc::async_notify_when_done` function as well. I will dig into it for the next release (2.3.0).

**Comment by 4eUeP** [comment]
It works. Btw, I have to call `grpc_context.stop` after `server.Shutdown` or it will hang forever in my environment.

**Comment by 4eUeP** [comment]
Another question: for a generic server, does this means for each request done, not for a client exit, I'll receive a notification?

```cpp
// request_loop
  grpc::GenericServerContext server_context;
  grpc::GenericServerAsyncReaderWriter reader_writer{&server_context};

  auto on_done = async_notify_when_done(grpc_context, server_context,
                                        asio::experimental::use_promise);
  const bool ok =
      co_await agrpc::request(service, server_context, reader_writer);

  auto method = server_context.method();
  std::cout << "-> " << method << std::endl;
// ...
```

I noticed that from the grpc documentation: "AsyncNotifyWhenDone has to be called before the rpc starts". So can I only register for a specific rpc? Which means for example`if rpc_method == "/Srv/should_notify" then async_notify_when_done else do_nothing`. Or only get notified when the client exit?

**Comment by Tradias** [comment]
> It works. Btw, I have to call grpc_context.stop after server.Shutdown or it will hang forever in my environment.

That is exactly what I was trying to describe by saying "GrpcContext will never run out of work after server shutdown". You can make it work though:

```c++
template <class RequestHandler>
asio::awaitable<void> request_loop(agrpc::GrpcContext& grpc_context, test::v1::Test::AsyncService& service,
                                   RequestHandler request_handler)
{
    grpc::ServerContext server_context;
    auto on_done = async_notify_when_done(grpc_context, server_context, asio::experimental::use_promise);
    asio::post(grpc_context,
               [&]
               {
                   // Discount the work of async_notify_when_done until the RPC is started
                   grpc_context.work_finished();
               });
    test::msg::Request request;
    grpc::ServerAsyncWriter<test::msg::Response> writer{&server_context};
    const bool ok = co_await agrpc::request(&test::v1::Test::AsyncService::RequestServerStreaming, service,
                                            server_context, request, writer);
    // Undo the discount
    grpc_context.work_started();
    if (!ok)
    {
        co_return;
    }
    asio::co_spawn(grpc_context, request_loop(grpc_context, service, request_handler), asio::detached);
    co_await request_handler(server_context, request, writer, std::move(on_done));
}
```

I am afraid for generic RPCs you will always have to set up async_notify_when_done and then simply ignore the promise if the method is not `"/Srv/should_notify"`.

> Or only get notified when the client exit?

After async_notify_when_done you can inspect `server_context.IsCancelled()` to see whether the client cancelled the RPC or it finished normally.

**Comment by Tradias** [comment]
I am also finishing up https://github.com/Tradias/asio-grpc/pull/52 which will avoid the aforementioned memory leak.

**Comment by Tradias** [comment]
[async_notify_when_done](https://tradias.github.io/asio-grpc/structagrpc_1_1detail_1_1_async_notfiy_when_done_fn.html) has been merged and will be released with 2.3.0.

**Comment by 4eUeP** [comment]
Sorry for the delayed response. Having tested `agrpc::notify_when_done` with a generic server, there seems to be no performance reduction I have worried about (Since for each unary call, I have to setup notify_when_done that is definitely not used).

Here is the result tested on my local machine with ghz:

```
// notify_when_done
-> 10B
  Requests/sec: 33586.30
-> 1KB
  Requests/sec: 27963.49
-> 10KB
  Requests/sec: 21373.20
-> 100KB
  Requests/sec: 5188.63
-> 500KB
  Requests/sec: 1180.22
-> 1MB
  Requests/sec: 605.20
-> 2MB
  Requests/sec: 285.98

// without
-> 10B
  Requests/sec: 30697.54
-> 1KB
  Requests/sec: 28996.35
-> 10KB
  Requests/sec: 23081.48
-> 100KB
  Requests/sec: 5035.29
-> 500KB
  Requests/sec: 1208.29
-> 1MB
  Requests/sec: 590.83
-> 2MB
  Requests/sec: 286.33
```

**Comment by Tradias** [comment]
That is good news, thanks for running the benchmark. I guess I should add a benchmark for a generic server to https://github.com/Tradias/grpc_bench anyways :).

**Comment by Tradias** [comment]
I have released [v2.3.0](https://github.com/Tradias/asio-grpc/releases/tag/v2.3.0), thereby closing this issue.

Do not hesitate to open a new issue if you have more questions or suggestions. I am also happy to chat about your haskell gRPC adaptation :).
